teps/0139-trusted-artifacts.md

```diff
@@ -0,0 +1,477 @@
+- This proposal does not discuss how to expose artifacts outside of a pipeline, even though it sets foundations that could be used to achieve that
+- This proposal does not discuss how to inject artifact as inputs to a pipeline, even though it sets foundations that could be used to achieve that. For instance, one could use a workspace preprovisioned with artifacts and use artifact type params as inputs for a pipeline
+
+### Use Cases
+
+- Extend the chain of trust across Tasks for provenance produced by Tekton Chains based on the `TaskRuns` and `PipelineRuns` executed by Tekton Pipeline
+
+### Requirements
+
+Support the following combinations:
+
+- One producer, one consumer
+- One producer, N consumers, including with write-one, read-many storage class
+- Many producers, one consumer
+- Many producers, many consumers, including with write-one, read-many storage class
+- Fail validation if workspaces (static/runtime) are not fit for artifacts
+- Fail execution if checksum validation fail, surface error to TaskRun/PipelineRun failure reason
+
+## Proposal
+
+In short the proposal is to:
+
+- extend parameters and result types to a new type `artifact`, an object type with a fixed schema
+- extend the `Pod` logic in the controller to inject hashing and checking steps when required
+- extend variable expansion to include `.data.path` for artifact parameters and results
+
+The same pipeline from demo, rewritten after implementation, is shown by this [demo pipeline](https://gist.github.com/afrittoli/7236be5fca524b752c221d2346497bb7):
+
+<details>
+  <summary>Demo Pipeline:</summary>
+
+```yaml=
+apiVersion: tekton.dev/v1
+kind: PipelineRun
+metadata:
+  generateName: trusted-artifacts-sugar
+spec:
+  pipelineSpec:
+    workspaces:
+      - name: artifactStorage  # In this example this is where we store artifacts
+        artifacts: true # this will result in failed validation if the workspace is bound to a readonly backend like a secret
+    tasks:
+      - name: producer
+        taskSpec:
+          results:
+            - name: aFileArtifact
+              type: artifact # inbuilt object schema (path, hash, type)
+              description: An artifact file
+            - name: aFolderArtifact
+              type: artifact # inbuilt object schema (path, hash, type)
+              description: An artifact folder
+          steps:
+            - name: produce-file
+              image: bash:latest
+              script: |
+                #!/usr/bin/env bash
+                # Produce some content. The result "data.path" will resolve to an empty dir
+                date +%s | tee "$(results.aFileArtifact.data.path)/afile.txt"
+                # The controller appends a step that calculates the hash, copies the data to
+                # the workspace marked for artifacs, builds the object result json,
```

Comment:
I wonder if we should provide more details about this `Artifact workspace`. 
- How large is it ?
- Is the size configurable by the service providers/end users? 
    - For example, on a multi-node cluster it might need to be accessible by the step that copies it to/from the user's emptyDir.
- Is it something that is automatically deleted by the controller at the end of the pipelineRun?

Reply:
In this proposal it's a regular workspace, so anything that can be bound to a workspace in write mode (i.e. no secrets or configmaps) is valid. The behaviour is the same as for other workspaces, there is no change proposed in the workspace implementation really.

Reply:
In terms of scheduling, the affinity assistant will continue to work the same way it does today.